### PR TITLE
deprecate `@pure` and make it just alias to `@assume_effects :foldable`

### PR DESCRIPTION
**Pull request by aviatesk**
This PR deprecates the `@pure` macro and removes all the internals that relate to the macro. It completes and replaces #48588.

### CODE REVIEW

File: base/compiler/abstractinterpretation.jl [hunk 1]
```diff
@@ -788,33 +784,6 @@ struct MethodCallResult
...
-    return method.pure
-end
-is_method_pure(match::MethodMatch) = is_method_pure(match.method, match.spec_types, match.sparams)
-
-function pure_eval_call(interp::AbstractInterpreter,
-    @nospecialize(f), applicable::Vector{Any}, arginfo::ArgInfo)
-    pure_eval_eligible(interp, f, applicable, arginfo) || return nothing
-    return _pure_eval_call(f, arginfo)
-end
 function _pure_eval_call(@nospecialize(f), arginfo::ArgInfo)
```
Comment (vchuravy) [inline base/compiler/abstractinterpretation.jl:787]: Add a comment that this is still being used for `abstract_call` for `typejoin`?

**Comment by vtjnash** [comment]
Please try to preserve and add Co-authored-by tags

File: base/compiler/abstractinterpretation.jl (hunk 1 above)
Comment (aviatesk) [reply base/compiler/abstractinterpretation.jl:787]: xref #48684

**Comment by aviatesk** [comment]
Whoops, I missed that I lost the tags. Thanks.

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/48682#issuecomment-1432595433) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/d38fb12_vs_75215fa/report.md).

**Comment by aviatesk** [comment]
No regressions. This should be ready to go.

**Comment by KristofferC** [comment]
-1 for the hard deprecation here. It is an internal that changes to another internal. Just put it in `deprecated.jl` without a deprecation and be done with it. No need to spam gazillions of warnings to normal users.

**Comment by oscardssmith** [comment]
the counterpoint is that 90% of uses of `@pure` were broken anyway.

**Comment by KristofferC** [comment]
How is that a counter point? Whose life is better now when we get a bazillion warnings from `precompile` for any package install? 

>  I'm hoping the result is that all the advanced people who run on master go around and fix all of the broken uses of @pure

No one is stopping anyone from doing that. Just don't spam my terminal with pages of warnings.

**Comment by maleadt** [comment]
This also broke Compat.jl 3.x, which quite a few packages still use.
https://github.com/JuliaLang/Compat.jl/blob/v3.46.0/src/Compat.jl#L282
IMO removals like this should have been preceded by a PkgEval run, and fixes to packages.

**Comment by martinholters** [comment]
Would that be solved by making the `if` in Compat a `@static if`? We could do that in a patch release of the v3 branch, I guess.

**Comment by maleadt** [comment]
Yeah, I think that should work.

**Comment by StefanKarpinski** [comment]
With the new definition, isn't it more likely that uses of `@pure` are now correct?

**Comment by oscardssmith** [comment]
They would be, but most uses of `@pure` were straight up broken. Eg `@pure f() = error()`. Almost all uses of `@pure` were pre 1.0 and are neither correct nor needed.

**Comment by aviatesk** [comment]
`Base.@assume_effects :foldable f() = error()` is valid :p I don't know much about `@pure` usages in the wild, but I guess they are not all correct even given the new definition.

**Pull request merged**